### PR TITLE
Added support for MPS Apple Silicon and quality of life improvements to the GUI

### PR DESCRIPTION
**Pull request by jjhickmon**
With increased MPS support from PyTorch, this feature will be necessary for users to run XMem on-device.
Provided is a brief summary of the changes:
- Updated code to use PyQt6
- Added checks for MPS instead of just CUDA
- Added code to correctly start the timeout sequence for the video player
- Tracking memory usage is not supported by PyTorch, so I patched a temporary fix where I just return their currently used memory, until PyTorch updates their MPS support
- Changed the shortcuts to be Cmd/Cntrl + (1 - N) and I also added a new widget to the GUI where users could manually select the object they wanted to select/edit. The keys 1 through N did not work if users were on Mac because of way Qt handles Mac shortcuts. Along with this, if users wanted to select more than 9 objects they previously could not.
- Added widget to export video of overlayed frames.

A few remaining issues include:
- Shortcuts such as the arrow keys are not working on Mac
- Accessing total memory is not supported by PyTorch for mps
- The eval.py file among others has not been fully updated to support mps. This pull request only includes code related to the GUI

### CODE REVIEW

**Comment by hkchengrex** [comment]
Hi @jjhickmon, thanks a lot for the contribution. I don't have time to look at it right now -- feel free to ping me if I didn't reply within a week.

File: interactive_demo.py [hunk 1]
```diff
@@ -64,32 +71,43 @@
             help='Resize the shorter side to this size. -1 to use original resolution. ')
     args = parser.parse_args()
 
+    # create temporary workspace if not specified
     config = vars(args)
     config['enable_long_term'] = True
     config['enable_long_term_count_usage'] = True
 
-    with torch.cuda.amp.autocast(enabled=not args.no_amp):
+    if config["workspace"] is None:
```
Comment (hkchengrex) [inline interactive_demo.py:79]: ResourceManager should have handled this. Am I missing something?

**Review by hkchengrex — CHANGES_REQUESTED** [review]
Thanks again for the contribution. I have left some comments that might be useful.

File: inference/interact/gui.py [hunk 2]
```diff
@@ -21,13 +21,14 @@
 
 import numpy as np
 import torch
+from torch import mps
 
-from PyQt5.QtWidgets import (QWidget, QApplication, QComboBox, QCheckBox,
+from PyQt6.QtWidgets import (QWidget, QApplication, QComboBox, QCheckBox,
```
[the remark this answers is not in the document]
Comment (jjhickmon) [reply inference/interact/gui.py:26]: Primarily switched to utilize the most up-to-date version of PyQt as a quality of life improvement, to ensure smooth support for future Python versions. An important note may be that PyQt6 only supports Python v3.6.1 and later, so even though it will support future Python versions, if it is needed to use older Python versions then I could revert this change.

File: eval.py [hunk 3]
```diff
@@ -218,7 +218,7 @@
                 prob = torch.flip(prob, dims=[-1])
 
             # Probability mask -> index mask
-            out_mask = torch.argmax(prob, dim=0)
+            out_mask = torch.max(prob, dim=0).indices
```
[the remark this answers is not in the document]
Comment (jjhickmon) [reply eval.py:221]: This was because of an PyTorch error on the MPS backend: https://github.com/pytorch/pytorch/issues/98191
Although it seems as though this issue has recently been resolved in the Nigtly build of PyTorch (but not the stable build 2.0.1) so should I amend this to use argmax as before?

File: inference/interact/gui.py [hunk 4]
```diff
@@ -622,6 +639,7 @@ def on_prev_frame(self):
         self.tl_slider.setValue(self.cursur)
 
     def on_next_frame(self):
+        print('next frame')
```
[the remark this answers is not in the document]
Comment (jjhickmon) [reply inference/interact/gui.py:642]: Correct, removed!

File: inference/interact/interactive_utils.py [hunk 5]
```diff
@@ -15,7 +15,7 @@ def image_to_torch(frame: np.ndarray, device='cuda'):
     return frame_norm, frame
 
 def torch_prob_to_numpy_mask(prob):
-    mask = torch.argmax(prob, dim=0)
+    mask = torch.max(prob, dim=0).indices
```
[the remark this answers is not in the document]
Comment (jjhickmon) [reply inference/interact/interactive_utils.py:18]: Described above

File: inference/interact/s2m_controller.py [hunk 6]
```diff
@@ -19,6 +19,7 @@ def __init__(self, s2m_net:S2M, num_objects, ignore_class, device='cuda:0'):
         self.device = device
 
     def interact(self, image, prev_mask, scr_mask):
+        print(self.device)
```
[the remark this answers is not in the document]
Comment (jjhickmon) [reply inference/interact/s2m_controller.py:22]: Yes, removed

File: interactive_demo.py (hunk 1 above)
Comment (jjhickmon) [reply interactive_demo.py:79]: Oh you are correct, I simply didn't notice that the ResourceManager kept a workspace variable. I will modify the PR to reflect these changes.

**Comment by jjhickmon** [comment]
Hi @hkchengrex thanks for the comments! I can raise the revision after you address the comment I left about torch.argmax. Along with this, while you were reviewing the PR I was able to fix the issue with the original shortcuts not working on Mac. Would you prefer I update this PR or raise a new one after? 

I also saw undo was partially implemented and I have a working implementation locally, but would you prefer the published code to not have the undo functionality?

**Comment by hkchengrex** [comment]
Hi, 

I have made some minor changes to the aesthetic (PyQt6 does not expand the draw area properly). I have also changed the video output format to mp4 and coded some backward compatibility for older PyTorch without MPS. I changed the "mask LCD" to "object dial", and changed it from a text box to a spin box for better input validation.
Can you test that it still works on Mac? 

Feel free to push any more fixes to the shortcuts.
As for the "undo" function, I left it out because it was too time-consuming to make bug-free. Pull requests for this function are definitely welcomed. I would expect a working undo function to go into each of the interaction types (e.g., we can undo click/scribble/free drawing) and restore states perfectly (i.e., if the user reset the mask with a click interaction and regretted it, they can always go back). The MiVOS version does not do this perfectly which is why I abandoned it here. That would be quite a lot of code changes orthogonal to the current one so it should be a different pull request.

**Comment by hkchengrex** [comment]
I'm merging because this adds good functionalities with no apparent bugs. Thank you for your contribution.
